hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,739 @@
+
+        const unsigned offset_front = addr - src_aligned;
+
+        memcpy(data_buf, tmpbuf + offset_front, data_size);
+    }
+
+    return 0;
+}
+
+int hal_exflash_erase_sector(uintptr_t start_addr, size_t num_sectors) {
+    return erase_common(start_addr, num_sectors, NRF_QSPI_ERASE_LEN_4KB);
+}
+
+int hal_exflash_erase_block(uintptr_t start_addr, size_t num_blocks) {
+    return erase_common(start_addr, num_blocks, NRF_QSPI_ERASE_LEN_64KB);
+}
+
+int hal_exflash_copy_sector(uintptr_t src_addr, uintptr_t dest_addr, size_t data_size) {
+    FlashLock lk;
+
+    uint8_t data_buf[FLASH_OPERATION_TEMP_BLOCK_SIZE] __attribute__((aligned(4)));
+    unsigned index = 0;
+    uint16_t copy_len;
+    uint16_t sector_num;
+
+    if ((src_addr % sFLASH_PAGESIZE) || (dest_addr % sFLASH_PAGESIZE) || !(IS_WORD_ALIGNED(data_size))) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+
+    // erase sectors
+    sector_num = CEIL_DIV(data_size, sFLASH_PAGESIZE);
+    CHECK(hal_exflash_erase_sector(dest_addr, sector_num));
+
+    // memory copy
+    for (index = 0; index < data_size; index += copy_len) {
+        copy_len = std::min((data_size - index), sizeof(data_buf));
+        CHECK(hal_exflash_read(src_addr + index, data_buf, copy_len));
+        CHECK(hal_exflash_write(dest_addr + index, data_buf, copy_len));
+    }
+
+    return 0;
+}
+
+int hal_exflash_read_special(hal_exflash_special_sector_t sp, uintptr_t addr, uint8_t* data_buf, size_t data_size) {
+    const hal_exflash_params_t* flash_params = hal_exflash_get_params(flash_type);
+    CHECK_TRUE(flash_params, SYSTEM_ERROR_NOT_FOUND);
+
+    CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + data_size <= flash_params->otp_size, SYSTEM_ERROR_OUT_OF_RANGE);
+    CHECK_TRUE(data_buf != nullptr, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int ret = 0;
+
+    FlashLock lk;
+
+    if (flash_type == HAL_QSPI_FLASH_TYPE_GD25WQ64E) {
+        ret = gd25_security_register_operation(gd25_secure_register_read, addr, data_buf, data_size);
+    } else {
+        ret = mx25_enter_secure_otp();
+        if (!ret) {
+            ret = hal_exflash_read(addr, data_buf, data_size);
+        }
+        mx25_exit_secure_otp(); 
+    }
+
+    return ret;
+}
+
+int hal_exflash_write_special(hal_exflash_special_sector_t sp, uintptr_t addr, const uint8_t* data_buf, size_t data_size) {
+    const hal_exflash_params_t* flash_params = hal_exflash_get_params(flash_type);
+    CHECK_TRUE(flash_params, SYSTEM_ERROR_NOT_FOUND);
+
+    CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + data_size <= flash_params->otp_size, SYSTEM_ERROR_OUT_OF_RANGE);
+    CHECK_TRUE(data_buf != nullptr, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int ret = 0;
+
+    FlashLock lk;
+
+    if (flash_type == HAL_QSPI_FLASH_TYPE_GD25WQ64E) {
+       ret = gd25_security_register_operation(gd25_secure_register_write, addr, (uint8_t *)data_buf, data_size);
```

Comment:
I had to drop the `const` in order to provide a common function signature for the common security register read/write function, but re-cast the data buffer back to const when passing it to the NRF SDK function that uses it to write the data